### PR TITLE
model ana-class100 | v1.80: Integrate ana-class100 models into updated processing pipeline

### PR DESCRIPTION
**Pull request by Slider0007**
- Adapt `ana-class100` models based on v1.71 to updated processing pipeline (github actions to train model)
- Adapt augmentation to be aligned with `ana-cont` model
  - Augmentation, most probable ones, incl.
      - White balance variances
      - Inverted images
- Retrained models based on updated image dataset (new images) and updated augmentation
- Finetune labeling of some images

---
Model name: `ana-class100_0180_...`
Model version: v1.80
Model comparison: [Overview](https://github.com/Slider0007/neural-network-analog-needle-readout/blob/update-ana-class100/models/README.md)

### CODE REVIEW

**Comment by Slider0007** [comment]
@haverland: Ich habe nun das ana-class100 model reaktiviert und in die die gleiche processing pipeline wie ana-cont integriert. Somit kann dies auch u.a. über die Github Actions prozessiert und ausgewertet werden (Hotwo: https://github.com/Slider0007/neural-network-analog-needle-readout/pull/4).

Basis: ana-class100 v1.71 + angepasstes Augmentation angelehnt an den Vorschlag, den ich hier (für ana-cont) bereits eingebracht hatte: https://github.com/jomjol/neural-network-analog-needle-readout/pull/33

Gerne reviewen und optimieren, sofern dir was auffällt.

Ich denke, es wäre schön, wenn wir zwei gleichwertige, aber dennoch verschiedene Ansätze (ana-class und ana-cont) hätten. Ggf. funktioniert beim einen das eine oder andere besser. Wie seht ihr das?

@jomjol: FYI

**Comment by haverland** [comment]
https://github.com/Slider0007/neural-network-analog-needle-readout/pull/11 auf deinem Branch fixed Überschriftsprobleme mit plot-function.

Erster Lauf ist recht vielversprechend. Gibt es einen Grund die Standard Learning-Rate von 1E-3 auf 5-E4 zu setzen?

**Comment by Slider0007** [comment]
@haverland 

Danke für den Fix, ist gemergt. Das mit den Nullen ist mir beim Testen zwar auch aufgefallen, hatte es aber wieder verdrängt

Zur Learning Rate: Ich hatte beim Recherchieren irgendwo gelesen, dass bei Adam eine niedrige Learning Rate teilweise hilfreich sein kann und es daher einfach ausprobiert. Die Accuracy ist bei meinen Trainingsversuchen dadurch etwas höher, aber nicht viel, 96+ zu 97+ mit der niedrigeren Rate. Die Trainingsdauer war für mich dabei untergeordnet. Sofern du denkst Standard wäre aus anderen Gründen geeigneter, können wir diese auch gerne wieder zurück stellen.

**Comment by haverland** [comment]
Ich kann das bestätigen. Zumal die Anzahl der Epochen dabei runter geht, was ein Overfitting reduziert. Ich probiere noch ReduceLROnPlateau die patience. S1 hat ja 10. S2 = 5. Ich vermute 5-7 dürfte auch die Anzahl der Epochen reduzieren.

Invertieren hat übrigens keinen Effekt. Das hatte ich schon vor 2 Jahren festgestellt und es deshalb nicht mehr verwendet. Der rote Zeiger ist ein guter Ansatzpunkt, wobei bei invertieren hier nur auf weiteren Neuronen die invertierte Kante gelernt wird.

**Comment by haverland** [comment]
Wenn es bedarf für invertierte Bilder gibt, ist das sicherlich ohne weiteres möglich. Das neuronale Netz wird dadurch nicht größer. Wie gesagt ist es nur die erste Schicht, die davon betroffen ist. Auf den Gedanken, dass jemand invertierte Bilder prozessieren will bin ich nur noch nicht gekommen. 
Die Erkennungsrate dürfte dann ähnlich den normalen Bildern sein. Ist eben nur eine Kantenerkennung. Farbe spielt noch eine Rolle bei der Erkennung des Zeigers. Der ist aktuell immer rot bzw. invertiert, wenn es benötigt wird.

**Comment by Slider0007** [comment]
@haverland:

Haben deine Untersuchungen zur Modellpflege noch was ergeben oder können wir die neue Version so erst einmal fixieren?

Als nächstes können wir dann noch der Vollständigkeit halber auf das image resizing mittels Mitchell-Filter switchen, obwohl meine ersten Untersuchungen gezeigt haben, dass der Unterschied bei den kleinen Bildgrößen fast vernachlässigbar ist. Dies würde aber dann beide Modelle (ana-cont und ana-class) gleichermaßen betreffen, da ab diesem PR für beide Modelle die gleiche pre-processing pipeline verwendet wird.

**Comment by haverland** [comment]
Bin zwar noch nicht ganz fertig, aber die Anpassungen an patience brachten keine Veränderung.

Insofern würde ich hier nichts ändern. 

Wenn du das resizing änderst, mag das beim Training etwas ändern. Nur auf dem ESP wird dann ein anderes Verfahren verwendet. Insofern wäre ich dafür hier das gleiche Verfahren zu verwenden, sonst sind die Aussagen über die Ergebnisse nicht mit dem Ergebnis auf dem ESP vergleichbar. Insbesondere beim Compare_Models, das ja extra die fertigen q.tflite vergleicht sollte das Verfahren vom ESP beibehalten werden.

**Comment by Slider0007** [comment]
@haverland 

> Wenn du das resizing änderst, mag das beim Training etwas ändern. Nur auf dem ESP wird dann ein anderes Verfahren verwendet. Insofern wäre ich dafür hier das gleiche Verfahren zu verwenden, sonst sind die Aussagen über die Ergebnisse nicht mit dem Ergebnis auf dem ESP vergleichbar. Insbesondere beim Compare_Models, das ja extra die fertigen q.tflite vergleicht sollte das Verfahren vom ESP beibehalten werden.

Bisher (alle? bisherigen Modelle vor diesem PR) wurde beim Resizing und Modell-Vergleich mit nearest neighbor prozessiert, was nicht dem eingesetzten Filter des ESP entspricht (Mitchell-Filter laut stb lib). Das Ziel wäre nun genau das was du voschlägst, dass wir für das pre-processing und den Modell-Vergleich auch den Mitchell Filter verwenden (was nach Analyse von Sybexx mit tensorflow image lib geht, aber nicht mit der pillow), um nun identisch zum ESP zu werden. Da dies aber beide Modelle betrifft, sollten wir zuerst diesen PR mergen und dann die Umstellung des resizing algos  für beide Modelle gleichermaßen machen.

**Comment by haverland** [comment]
OK. Ich hatte nur @jomjol s Aussage im Kopf. Wenn ihr das geprüft hat passt es ja.
